File: Go/Go_100-999/556.go
// Go
// leetcode 556
// https://leetcode.cn/problems/next-greater-element-iii/

func nextGreaterElement(n int) int {
	arr, pre := make([]int, 0, 2), n
	for n > 0 {
		arr = append(arr, n % 10)
		n /= 10
	}
	n = len(arr)
	p := n - 1
	reverse(arr, 0, p)
	for p >= 1 && arr[p-1] >= arr[p] {
		p--
	}

	if p != 0 {
		q, tmp := p, 0
		for q + 1 < n && arr[q + 1] > arr[p - 1] {
			q++
		}
		tmp = arr[p - 1]
		arr[p - 1] = arr[q]
		arr[q] = tmp
		reverse(arr, p, n - 1)
		reverse(arr, 0, n - 1) 
	}
	var a int64
	for i := n - 1; i >= 0; i-- {
		a = a * 10 + int64(arr[i])
	}
	if a > 0b1111111111111111111111111111111 {
		return -1
	}
	res := int(a)
	if(res <= pre) {
		return -1
	}
	return res
}

func reverse(arr []int, l, r int) {
	var tmp int
	for l < r {
		tmp = arr[l]
		arr[l] = arr[r]
		arr[r] = tmp
		l++
		r--
	}
}

File: cpp/cpp_100-999/556.cc
// C++
// leetcode 556
// https://leetcode.cn/problems/next-greater-element-iii/

class Solution {
public:
    int nextGreaterElement(int n) {
    	int nn = n;
    	vector<int> nums;
    	while (n) {
    		nums.push_back(n % 10);
    		n /= 10;
    	}
    	n = nums.size();
    	int p = n - 1;
    	reverse(nums, 0, p);

    	while(p >= 1 && nums[p - 1] >= nums[p])
            --p;
        if(p == 0) {
            reverse(nums, 0, n - 1);
        } else {
            int q = p;
            while(q + 1 < n && nums[q + 1] > nums[p - 1])
                ++q;
            swap(nums[p - 1], nums[q]);
            reverse(nums, p, n - 1);
        }


        long res = 0;
        for(int i = 0; i < n; i++) {
        	res = res * 10 + nums[i];
        }
        // cout<<res<<endl;
        if (res > (long) INT_MAX)
        	return -1;
        n = (int)res;
        if(n <= nn)
        	return -1;
        return n;
    }

    void reverse(vector<int> &arr, int l, int r) {
    	int tmp;
    	while(l < r) {
    		tmp = arr[l];
    		arr[l] = arr[r];
    		arr[r] = tmp;
    		l++;
    		r--;
    	}
    }
};